File: app/src/main/java/info/xudshen/android/playground/view/adapter/SampleItemAdapter.java
package info.xudshen.android.playground.view.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import info.xudshen.android.playground.R;
import info.xudshen.android.playground.data.SampleItemDataSource;
import info.xudshen.android.playground.model.SampleItemModel;

/**
 * Created by xudong on 2017/1/11.
 */

public class SampleItemAdapter extends RecyclerView.Adapter<SampleItemAdapter.ViewHolder> {
    private OnClickListener onClickListener;

    public SampleItemAdapter() {
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(viewType, parent, false));
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        SampleItemModel itemModel = SampleItemDataSource.DATA.get(position);
        holder.title.setText(itemModel.getTitle());
    }

    @Override
    public int getItemCount() {
        return SampleItemDataSource.DATA.size();
    }

    @Override
    public int getItemViewType(int position) {
        return R.layout.layout_sample_item;
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.section_title)
        TextView title;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (onClickListener != null) {
                        onClickListener.onClick(view, ViewHolder.this, getAdapterPosition(),
                                SampleItemDataSource.DATA.get(getAdapterPosition()));
                    }
                }
            });
        }
    }

    public interface OnClickListener {
        void onClick(View view, ViewHolder viewHolder, int position, SampleItemModel itemModel);
    }

    public OnClickListener getOnClickListener() {
        return onClickListener;
    }

    public void setOnClickListener(OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }
}

File: app/src/main/java/info/xudshen/android/playground/view/adapter/SampleItemAdapter.kt
package info.xudshen.android.playground.view.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.nostra13.universalimageloader.core.ImageLoader
import info.xudshen.android.playground.R
import info.xudshen.android.playground.data.SampleItemDataSource
import info.xudshen.android.playground.model.SampleItemModel
import kotlinx.android.synthetic.main.layout_sample_item.view.*

/**
 * Created by xudong on 2017/1/11.
 */
class SampleItemAdapter : RecyclerView.Adapter<SampleItemAdapter.ViewHolder>() {
    var onClickListener: ((view: View, viewHolder: ViewHolder,
                           position: Int, itemModel: SampleItemModel?) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
            ViewHolder(LayoutInflater.from(parent.context).inflate(viewType, parent, false))

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val itemModel = SampleItemDataSource.DATA[position]
        holder.itemView.section_title.text = itemModel.title
        ImageLoader.getInstance().displayImage("drawable://${itemModel.cover}", holder.itemView.section_cover)
    }

    override fun getItemCount(): Int = SampleItemDataSource.DATA.size

    override fun getItemViewType(position: Int): Int = R.layout.layout_sample_item

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            itemView.setOnClickListener {
                onClickListener?.invoke(it, this@ViewHolder, adapterPosition,
                        SampleItemDataSource.DATA.elementAtOrNull(adapterPosition))
            }
        }
    }
}
